probe-rs/src/probe/stlink/mod.rs

```diff
@@ -834,6 +1101,632 @@ impl From<StlinkError> for ProbeCreationError {
     }
 }
 
+#[derive(Debug)]
+struct StlinkArmDebug {
+    probe: Box<STLink<STLinkUSBDevice>>,
+    state: ArmCommunicationInterfaceState,
+}
+
+impl StlinkArmDebug {
+    fn new(probe: Box<STLink<STLinkUSBDevice>>) -> Result<Self, DebugProbeError> {
+        let state = ArmCommunicationInterfaceState::new();
+
+        /* determine the number and type of available APs */
+
+        let mut interface = Self { probe, state };
+
+        for ap in valid_access_ports(&mut interface) {
+            let ap_state = interface.read_ap_information(ap)?;
+
+            log::debug!("AP {}: {:?}", ap.port_number(), ap_state);
+
+            interface.state.ap_information.push(ap_state);
+        }
+
+        Ok(interface)
+    }
+
+    /// Read information about an AP from its registers.
+    ///
+    /// This reads the IDR register of the AP, and parses
+    /// further AP specific information based on its class.
+    ///
+    /// Currently, AP specific information is read for Memory APs.
+    fn read_ap_information(
+        &mut self,
+        access_port: GenericAP,
+    ) -> Result<ApInformation, DebugProbeError> {
+        let idr = self.read_ap_register(access_port, IDR::default())?;
+
+        if idr.CLASS == APClass::MEMAP {
+            let access_port: MemoryAP = access_port.into();
+
+            let base_register = self.read_ap_register(access_port, BASE::default())?;
+
+            let mut base_address = if BaseaddrFormat::ADIv5 == base_register.Format {
+                let base2 = self.read_ap_register(access_port, BASE2::default())?;
+
+                u64::from(base2.BASEADDR) << 32
+            } else {
+                0
+            };
+            base_address |= u64::from(base_register.BASEADDR << 12);
+
+            log::debug!(
+                "AP {} - BaseAddress: {:#08x}",
+                access_port.port_number(),
+                base_address
+            );
+
+            // Ensure if we can access this AP
+            let status = self.read_ap_register(access_port, CSW::default())?;
+
+            log::debug!("AP {} - {:?}", access_port.port_number(), status);
+
+            // TODO: Figure out how to handle this for STM32,
```

Comment:
I think STMs can only ever do 32bit accesses but I might be mistaken :) It's certainly good to leave this comment here.

Reply:
They offer a dedicated API for 8 bit access, which would be a bit pointless if they don't supoort it. At least when I try it, the 8 bit accesses do work. Of course, it might be that they just get translated to 32 bit accesses internally. Would be interesting to try it on a chip where we know that only 32 bit access is supported.